src/BoundaryConditions/flat_extrapolation_open_boundary_matching_scheme.jl

```diff
@@ -0,0 +1,108 @@
+using Oceananigans.Grids: xspacings, yspacings, zspacings
+
+"""
+    FlatExtrapolation
+
+Zero gradient perpendicular velocity boundary condition.
+
+For constant spacing:
+```math
+f′(xᵢ) ≈ f′(xᵢ₋₁) + f′′(xᵢ₋₁)(xᵢ₋₁ - xᵢ) + O(Δx²) = f′(xᵢ₋₁) + f′′(xᵢ₋₁)Δx + O(Δx²),
+
+f′(xᵢ₋₁) ≈ (f(xᵢ) - f(xᵢ₋₂)) / 2Δx
+
+f′(xᵢ) ≈ (f(xᵢ) - f(xᵢ₋₂)) / 2Δx + O(Δx) = 0 ∴ f(xᵢ) ≈ f(xᵢ₋₂) + O(Δx)
+
+f′′(xᵢ₋₁) ≈ (f′(xᵢ) - f′(xᵢ₋₂)) / 2Δx = - f′(xᵢ₋₂) / 2Δx ≈ - (f(xᵢ₋₁) - f(xᵢ₋₃)) / (2Δx)²
+
+∴ f(xᵢ) ≈ f(xᵢ₋₂) - (f(xᵢ₋₁) - f(xᵢ₋₃))/2 + O(Δx²)
+```
+
+When the grid spacing is not constant it works out that the factor of 1/2 changes to 
+``Δx₋₁/(Δx₋₂ + Δx₋₃)`` instead.
+"""
+struct FlatExtrapolation{FT}
+    relaxation_timescale :: FT
+end
+
+const FEOBC = BoundaryCondition{<:Open{<:FlatExtrapolation}}
+
+function FlatExtrapolationOpenBoundaryCondition(val = nothing; relaxation_timescale = Inf, kwargs...)
+    classification = Open(FlatExtrapolation(relaxation_timescale))
+    
+    return BoundaryCondition(classification, val; kwargs...)
+end
+
+@inline relax(l, m, c, bc, grid, clock, model_fields) =
+    c + ifelse(isnothing(bc.condition)||!isfinite(clock.last_stage_Δt), 0,
+        (getbc(bc, l, m, grid, clock, model_fields) - c) * min(1, clock.last_stage_Δt / bc.classification.matching_scheme.relaxation_timescale))
+
+@inline spacing_factor(args...) = 1/2
+@inline spacing_factor(Δ::AbstractArray, ::Val{:right}) = @inbounds Δ[end] / (Δ[end-1] + Δ[end-2])
+@inline spacing_factor(Δ::AbstractArray, ::Val{:left})  = @inbounds Δ[1]   / (Δ[2]     + Δ[3]    )
+
+@inline function _fill_west_open_halo!(j, k, grid, c, bc::FEOBC, loc, clock, model_fields)
+    Δx = xspacings(grid, Center(), Center(), Center())
+
+    unrelaxed = @inbounds c[3, j, k] - (c[2, j, k] - c[4, j, k]) * spacing_factor(Δx, Val(:left))
```

Comment:
what does `unrelaxed` mean?

Reply:
`xspacings` should not be called within a kernel. You need to use `xspacing`. (In the future `xspacings` may itself launch a kernel.

Reply:
I suggest writing six functions of the form `left_xspacing(grid)` and `right_xspacing(grid)`.

Reply:
```julia
const c = Center()

@inline function left_xspacing_factor(j, k, grid)
    Δx1 = xspacings(1, j, k, grid, c, c, c)
    Δx2 = xspacings(2, j, k, grid, c, c, c)
    Δx3 = xspacings(3, j, k, grid, c, c, c)
    return Δx1 / (Δx2 + Δx3)
end
```

and so on

Reply:
Given that I'm going to have to write them 6 times couldn't I just do it inline in the functions?